### DISCUSSION TITLE
Specify some behavior left undefined by C

### DISCUSSION THREAD

**Discussion by mdumitru** [seahorn/seahorn#426] [Q&A]
Hello! I am very new to SeaHorn and I was wondering whether SeaHorn is tied tightly to standard-C semantics, or if there is any way to specify a concrete manifestation of behaviors C leaves undefined (e.g. signed integer overflow).

My question is motivated by examples such as this:

```c
unsigned x = nd();
unsigned z;
assume(x > 0);
z = x + 1;
sassert(z > x);
```

Here, SeaHorn figures out that an overflow is possible if `x` is `UINT_MAX - 1`. However, changing the variable types to signed integers yields an `unsat` answer - fair enough, as signed integer overflow is undefined behavior.
But is there a way to inform SeaHorn that, for this program, signed integer overflow works with the x86 semantics, for example?

**Comment by caballa** [comment]
First, SeaHorn analyzes LLVM bitcode  and thus, LLVM semantics is the reference semantics instead of C semantics.
Second, SeaHorn has different modes that models the LLVM semantics at different level of abstractions with different correctness assumptions. The command `sea pf` models integers more loosely by using signed mathematical integers. This is fine as long as the program does not have integer overflows. If you want to reason about integer overflows then the command `sea bpf` models integers as machine integers. 

In your example, you get `unsat` because probably you are using `sea pf` which uses mathematical integers and thus, `x+1 > x` is always true and therefore the assertion holds and SeaHorn returns `unsat`. If you use `sea bpf` then you should get `sat` because the assertion does not hold if integer overflow.

**Reply by mdumitru** [reply]
Thank you for the swift reply!

I am using `bpf` as an argument and still get an `unsat`. The output contains this warning:

`WARNING: no assertion was found so either program does not have assertions or frontend discharged them.`

I suspected that `clang` somehow optimizes out the assertion when compiling the bitcode; but I broke down the actions of the `sea` script (forced an `-O0` on clang) and inspected the resulting bitcode, which seems to contain the parts relevant to the `sassert`. So I'm not sure what's going on.

**Reply by caballa** [reply]
To see the actual bitcode being analyzed use the option --oll=prog.ll

Even if clang is run with -O0 SeaHorn still optimizes the bitcode, so the bitcode you are looking at is not the final analyzed bitcode

**Reply by mdumitru** [reply]
Thank you! I got the actual bitcode under analysis and, indeed, assertions seem to have disappeared. But I am lost as to how to prevent these optimizations. Using `bpf` did not help, unfortunately.

**Reply by caballa** [reply]
Yes, the `bpf` mode performs even more optimizations because it statically unrolls loops which allows more optimizations.
In general, it's not a good idea to try to disable LLVM optimizations. 
But I'm missing the point, why do you want to avoid removing trivial assertions from LLVM programs?

**Reply by mdumitru** [reply]
I'm trying to perform some analysis of how the code in question would work under concrete circumstances; for example, compiled and run on an `x86` platform. Then, having warp-around semantics for signed integer overflow, I want SeaHorn to produce the answer `sat` for the following code and output the counterexample `INT_MAX` as a value for `x`:

```c
int x = nd(), z;
z = x + 1;
sassert(z > x);
```

**Reply by caballa** [reply]
Then two things:
1) If clang or llvm detects undefined behavior (UB) then it's up to them to deal with it by performing any transformation they want to apply. SeaHorn won't try to stop that and it shouldn't  because SeaHorn is a LLVM analyzer and not a C analyzer. 
2) However, this doesn't mean that you can still play with small examples that have UB as long as you hide enough that UB so that clang/llvm cannot detect it.

Since you want to reason about integer overflow you should use `bpf`. As I said, this mode performs even more transformation so hide UB to clang/llvm can be harder but it's possible. One possibility is to make sure that all operations are in memory  (i.e., use pointers) rather than scalar integers in your example. clang/llvm is very limited with memory. But again, LLVM has optimizations that try to lower memory operations to registers. One way to avoid these optimizations is to pass  the pointers that contain `x` and `z` to external functions (defined by you). In that way, LLVM cannot optimize since the external function might modify the pointer's content.

**Reply by mdumitru** [reply]
I've tried unsuccessfully to hide the UB from clang; but I've just revisited the problem and I think I was mistaken about its source. The code generated by `--oll` does have the assert removed (hence the no-assertions warning). But the executable generated with the `exe` option and the `--cex` counterexample has the assertion still there, it's just that the `nd` implementation returns `0`.

But if I don't nondeterministically initialize `x`, but rather give it a concrete value:

```c
int x = INT_MAX, z;
z = x + 1;
sassert(z > x);
```

SeaHorn produces a `sat` answer (and the executable later generated with the `exe` option does call `__VERIFIER_error`). However, this is still UB and I would've thought it suffers from the same problem as the previous example. I'm not sure what I'm missing.
